Source/Core/Core/PowerPC/MMU.h

```diff
@@ -15,14 +15,55 @@ namespace PowerPC
 {
 // Routines for debugger UI, cheats, etc. to access emulated memory from the
 // perspective of the CPU.  Not for use by core emulation routines.
-// Use "Host_" prefix.
+// Use "Host" prefix.
+
+enum class RequestedAddressSpace
+{
+  Effective,  // whatever the current MMU state is
+  Physical,   // as if the MMU was turned off
+  Virtual,    // specifically want MMU turned on, returns std::nullopt if off
+};
+
 u8 HostRead_U8(u32 address);
 u16 HostRead_U16(u32 address);
 u32 HostRead_U32(u32 address);
 u64 HostRead_U64(u32 address);
 float HostRead_F32(u32 address);
 double HostRead_F64(u32 address);
 u32 HostRead_Instruction(u32 address);
+std::string HostGetString(u32 address, size_t size = 0);
+
+template <typename T>
+struct TryReadResult
+{
+  bool success;
+  bool translated;
+  T value;
+
+  TryReadResult() : success(false) {}
+  TryReadResult(bool translated_, T value_) : success(true), translated(translated_), value(value_)
```

Comment:
I believe you want `value(std::forward(value_))`

Also your function definition should be `T&&`.  Otherwise your string moves are actually a copy

Reply:
This doesn't work, it prevents use of constant values as parameters, as far as I can tell. I just added an `std::move()` for the value, should be good enough in practice...

Reply:
> This doesn't work, it prevents use of constant values as parameters, as far as I can tell.

It should, it's perfect forwarding so const parameters would just get passed in as if you had a normal constructor...

> I just added an std::move() for the value, should be good enough in practice...

NBD but I believe that will still incur a copy in some cases

Reply:
~~~
  TryReadResult(bool translated_, T&& value_)
      : success(true), translated(translated_), value(std::forward(value_))
  {
  }
~~~

~~~
    const u32 value = ReadFromHardware<XCheckTLBFlag::OpcodeNoException, u32, true>(address);
    return TryReadResult<u32>(false, value);
~~~

~~~
2>Source\Core\Core\PowerPC\MMU.cpp(444,43): error C2440: '<function-style-cast>': cannot convert from 'initializer list' to 'PowerPC::TryReadResult<u32>'
2>Source\Core\Core\PowerPC\MMU.cpp(444,43): error C2440:     return TryReadResult<u32>(false, value);
2>Source\Core\Core\PowerPC\MMU.cpp(444,43): error C2440:                                           ^
2>Source\Core\Core\PowerPC\MMU.cpp(444,43): message : No constructor could take the source type, or constructor overload resolution was ambiguous
2>Source\Core\Core\PowerPC\MMU.cpp(444,43): message :     return TryReadResult<u32>(false, value);
2>Source\Core\Core\PowerPC\MMU.cpp(444,43): message :                                           ^
~~~

Reply:
Huh, I guess I always used a templated method in this scenario (where `const` would be forwarded), not a templated class, so I didn't notice.

Here's a variation of your current solution and should handle both the move and const variants with no copies:

```
  TryReadResult(bool translated_, T&& value_)
      : success(true), translated(translated_), value(std::move(value_))
  {
  }
  TryReadResult(bool translated_, const T& value_)
      : success(true), translated(translated_), value(value_)
  {
  }
```

Reply:
Did you mean std::move() in that first constructor? Or should that be std::forward()?